pkg/frame/native.go

```diff
@@ -1,62 +1,195 @@
+	//  16               []uint16
+	//  32               []uint32
+	RawDataSlice() any
+	// Equals returns true if this INativeFrame exactly equals the provided
+	// INativeFrame. This checks every pixel value, so may be expensive.
+	// In the future we may compute a one time hash during construction to make
+	// this less expensive in the future if called multiple time.
+	Equals(frame INativeFrame) bool
+	CommonFrame
+}
+
 // NativeFrame represents a native image frame
-type NativeFrame struct {
-	// Data is a slice of pixels, where each pixel can have multiple values
-	Data          [][]int
-	Rows          int
-	Cols          int
-	BitsPerSample int
+type NativeFrame[I constraints.Integer] struct {
+	// RawData is a slice of pixel values. For each pixel, each sample for the
+	// pixel is unrolled per pixel. For example, consider 2 pixels that have 3
+	// samples per Pixel: [[1,2,3], [4,5,6]]. This would be unrolled like
+	// [1,2,3,4,5,6]. The pixels themselves are arranged in row order, so the
+	// first row of pixels would be unrolled in order, followed by the next row,
+	// and so on in this flattened array.
+	// A flattened slice is used instead of a nested 2D slice because there is
+	// significant overhead to creating nested slices in Go discussed here:
+	// https://github.com/suyashkumar/dicom/issues/161#issuecomment-2143627792.
+	RawData                 []I
+	InternalSamplesPerPixel int
+	InternalRows            int
+	InternalCols            int
+	InternalBitsPerSample   int
+}
+
+// NewNativeFrame creates a new NativeFrame[I] given the input parameters. It
+// initializes the NativeFrame's internal RawData slice based on pixelsPerFrame
+// and samplesPerPixel.
+func NewNativeFrame[I constraints.Integer](bitsPerSample, rows, cols, pixelsPerFrame, samplesPerPixel int) *NativeFrame[I] {
+	return &NativeFrame[I]{
+		InternalBitsPerSample:   bitsPerSample,
+		InternalRows:            rows,
+		InternalCols:            cols,
+		RawData:                 make([]I, pixelsPerFrame*samplesPerPixel),
+		InternalSamplesPerPixel: samplesPerPixel,
+	}
 }
 
+// Rows returns the number of rows in this frame (which is the max y dimension).
+func (n *NativeFrame[I]) Rows() int { return n.InternalRows }
+
+// Cols returns the number of columns in this frame (which is the max x
+// dimension).
+func (n *NativeFrame[I]) Cols() int { return n.InternalCols }
+
+// BitsPerSample returns the bits per sample.
+func (n *NativeFrame[I]) BitsPerSample() int { return n.InternalBitsPerSample }
+
+// SamplesPerPixel returns the samples per pixel.
+func (n *NativeFrame[I]) SamplesPerPixel() int { return n.InternalSamplesPerPixel }
+
+// GetPixel returns the samples (as a slice) for the pixel at (x, y).
+// The coordinate system of the image starts with (0, 0) in the upper left
+// corner of the image, with X increasing to the right, and Y increasing
+// down:
+//
+//	0 -------▶ X
+//	|
+//	|
+//	▼
+//	Y
+func (n *NativeFrame[I]) GetPixel(x, y int) ([]int, error) {
```

Comment:
Can we have an API to manipulate the pixel data?

Smt like `SetPixel(x, y int, value []int)`

Reply:
Sure, sgtm! I can add that in, or feel free to send a PR if you'd like. Note that iiuc editing the data in the returned `RawDataSlice()` will also edit the data in the frame.